src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,137 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (head == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size == index) {
+            add(value);
+        } else {
+            Node<T> nodeByIndex = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            newNode.next.prev = newNode;
+            if (index != 0) {
+                newNode.prev.next = newNode;
+            } else {
+                head = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        T oldValue = nodeByIndex.value;
+        nodeByIndex.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        T removedValue = nodeByIndex.value;
+        removeNode(nodeByIndex);
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        while (node != null) {
+            if (object == node.value || object != null && object.equals(node.value)) {
+                removeNode(node);
+                size--;
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException();
```

Comment:
not informative exception message
```suggestion
            throw new ArrayIndexOutOfBoundsException("Invalid index " + index + " for size " + size);
```